File: BTVN2/bai5.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>
#include <algorithm>
using namespace std;
int main() {
    int v, e, n;
    cin >> v >> e >> n;
    vector<string> vertices(v);
    for(int i = 0; i < v; i++)
        cin >> vertices[i];
    map<string, vector<string>> graph;
    for(int i = 0; i < e; i++) {
        string u, v;
        cin >> u >> v;
        graph[u];
        graph[v];
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    vector<int> colors(v);
    for(int i = 0; i < v; i++)
        cin >> colors[i];
    map<string, int> color_map;
    set<int> used_colors;
    for(int i = 0; i < v; i++) {
        color_map[vertices[i]] = colors[i];
        if(colors[i] != -1)
            used_colors.insert(colors[i]);
    }
    vector<string> ques(n);
    for(int i = 0; i < n; i++)
        cin >> ques[i];
    for(auto & i : ques) {
        if(color_map[i] != -1) {
            bool flag = true;
            for(auto & j : graph[i]) {
                if(color_map[j] == color_map[i]) {
                    flag = false;
                    break;
                }
            }
            if(flag) {
                cout << "TRUE" << endl;
                continue;
            }
        }
        for(auto & color : used_colors) {
            bool flag = true;
            for(auto & j : graph[i]) {
                if(color_map[j] == color) {
                    flag = false;
                    break;
                }
            }
            if(flag) {
                color_map[i] = color;
                cout << color << endl;
                break;
            }
        }
    }
    cout<<graph.size() << endl;
    return 0;
}

File: BTVN2/bai5.py
def main():
    v, e, n = map(int, input().split())
    vertices = input().split()
    graph = {vertex: [] for vertex in vertices}
    for _ in range(e):
        u, i = map(str, input().split())
        graph[u].append(i)
        graph[i].append(u)
    colors = list(map(int, input().split()))
    color_map = {vertex: -1 for vertex in vertices}
    used_colors = set()
    for vertex, color in zip(vertices, colors):
        color_map[vertex] = color
        if color != -1:
            used_colors.add(color)
    ques = input().split()
    for i in ques:
        if color_map[i] != -1 and all(color_map[j] != color_map[i] for j in graph[i]):
            print("TRUE")
        else:
            for color in sorted(used_colors):
                if all(color_map[j] != color for j in graph[i]):
                    color_map[i] = color
                    print(color)
                    break
            else:
                new_color = 0
                while new_color in used_colors:
                    new_color += 1
                color_map[i] = new_color
                used_colors.add(new_color)
                print(new_color)

if __name__ == "__main__":
    main()